### PR TITLE
[nav] Document code

### PR DESCRIPTION
**Pull request by qhdwight**
- Added a bunch of comments to code
- Cleaned up include headers
- Made code more consistent, formatted with .clang-format

### CODE REVIEW

File: jetson/nav/README.md [hunk 1]
```diff
@@ -1,156 +1,203 @@
...
+correct ID.
 
 ---
 
-<!----------------------------- Search ----------------------------->
-## Search
-Similar to the `gate_search/` folder, this folder for search logic contains a `searchStateMachine` object and files to define the waypoints for different types of searches. First we follow a square spiral outwards with points generated in spiralOutSearch.cpp, then if the search completes and the target is not found, we will move onto trying the lawnmower search and the spiral in search.
+## Gate Search
+
+[`gateStateMachine.cpp`](./gate_search/gateStateMachine.cpp)
```
Comment (cgiger00) [inline jetson/nav/README.md:84]: should this be the same size/bolded like the other file names?

File: jetson/nav/README.md [hunk 2]
```diff
@@ -1,156 +1,203 @@
...
 
 ---
 
-<!----------------------------- Vars and Utils ----------------------------->
-## Variables and Utilities
+## Utilities
 
-#### Nav State
-The nav state specifies what state of the state machine we are in. It is implemented as an enum (a C++ type) where named states are associated with a number behind the scenes.
+[`utilities.cpp`](./utilities.cpp) Misc. math, conversions between GPS and cartesian coordinates \
```
Comment (cgiger00) [inline jetson/nav/README.md:101]: same comment as above

File: jetson/nav/README.md [hunk 3]
```diff
@@ -1,156 +1,203 @@
...
 
-#### Auton State
-This is just a boolean hidden as a type called AutonState and it tells us if Auton is on or off. It is read from LCMs published by the GUI or simulator.
+---
 
-##### `utilities.cpp`
-Contains functions used commonly throughout auton code. 
+## State Diagram
 
+![image](https://user-images.githubusercontent.com/20666629/170621044-2055089d-cba9-4164-8472-e669dfc1ee66.png)
```
Comment (cgiger00) [inline jetson/nav/README.md:110]: can you add also auton being enabled in going from off to drive_waypoints/done

Comment (cgiger00) [reply jetson/nav/README.md:110]: can you add also in seeing posts/gates it's 2 posts with the correct ID, ik earlier gate was optimized to see just two posts but that was changed bc it was causing incorrect behavior

File: jetson/nav/README.md [hunk 4]
```diff
@@ -1,156 +1,203 @@
...
 
 ---
 
-<!----------------------------- LCMs ----------------------------->
-## LCM Channels Publishing/Subscribed To 
+## LCM Channels Publishing/Subscribed To
+
 **Target [subscriber]** \
```
Comment (cgiger00) [inline jetson/nav/README.md:116]: **Enable [subscriber]**

File: jetson/nav/README.md [hunk 5]
```diff
@@ -1,156 +1,203 @@
...
-Messages: [ Course.lcm ](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Course.lcm) “/course” \
+Messages: [Course.lcm](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Course.lcm) “/course” \
 Publishers: simulators/nav, base_station/gui \
 Subscribers: jetson/nav
 
 **Obstacle [subscriber]** \
-Messages: [ Obstacle.lcm ](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Obstacle.lcm) “/obstacle” \
+Messages: [Obstacle.lcm](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Obstacle.lcm)
+“/obstacle” \
 Publishers: jetson/percep \
```
Comment (cgiger00) [inline jetson/nav/README.md:130]: sim also publishes right?

File: jetson/nav/README.md [hunk 6]
```diff
@@ -1,156 +1,203 @@
...
 Subscribers: jetson/nav
 
 **Odometry [subscriber]** \
-Messages: [ Odometry.lcm ](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Odometry.lcm) “/odometry” \
+Messages: [Odometry.lcm](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Odometry.lcm)
+“/odometry” \
 Publishers: jetson/filter \
```
Comment (cgiger00) [inline jetson/nav/README.md:136]: & sim?

File: jetson/nav/README.md [hunk 7]
```diff
@@ -1,156 +1,203 @@
...
 Subscribers: jetson/nav
 
-**ZED Gimbal Data [subscriber]** \
-Messages: [ ZedGimablPosition.lcm ](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/TargetList.lcm) “/zed_gimbal_data” \
-Publishers: simulators/nav, raspi/zed_gimbal, jetson/nav (TODO) \
-Subscribers: jetson/nav 
-
 **Joystick [publisher]** \
-Messages: [ Joystick.lcm ](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Joystick.lcm) “/autonomous” \
+Messages: [Joystick.lcm](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/Joystick.lcm)
```
Comment (cgiger00) [inline jetson/nav/README.md:146]: this isn't a thing anymore right? it's AutonDriveControl

File: jetson/nav/README.md [hunk 8]
```diff
@@ -1,156 +1,203 @@
...
+“/autonomous” \
 Publishers: jetson/nav \
 Subscribers: jetson/teleop, simulators/nav, base_station/gui
 
 **NavStatus [publisher]** \
-Messages: [ NavStatus.lcm ](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/NavStatus.lcm) “/nav_status” \
+Messages: [NavStatus.lcm](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/NavStatus.lcm)
+“/nav_status” \
```
Comment (cgiger00) [inline jetson/nav/README.md:153]: science bridge doesn't subscribe anymore, that was w the old led interface

File: jetson/nav/README.md [hunk 9]
```diff
@@ -1,156 +1,203 @@
...
 Publishers: jetson/nav \
 Subscribers: simulators/nav, base_station/gui, jetson/science_bridge
 
+**NavStatus [publisher]** \
+Messages: [ProjectedPoints.lcm](https://github.com/umrover/mrover-workspace/blob/master/rover_msgs/ProjectedPoints.lcm)
+“/projected_points” \
+Publishers: jetson/nav
```
Comment (cgiger00) [inline jetson/nav/README.md:160]: base_station/gui subscribes right?

File: jetson/nav/README.md [hunk 10]
```diff
@@ -1,156 +1,203 @@
...
 
-If desired, you can run a fourth terminal for debugging purposes via LCM messages. To do so, make sure you have another terminal, and starting in the `mrover-workspace` directory, run `vagrant ssh`. Once we are ssh'ed into the virtual machine, run `$./jarvis build lcm_tools/echo` to build the echo tool for LCMs. This will return the messages that are being communicated between publishers and subscribers. To run, enter the command `$./jarvis exec lcm_tools/echo TYPE_NAME CHANNEL` to echo the specified LCM and channel. (These are described in our LCM section and ICDs on the Drive)
+Run the command `./jarvis build simulators/nav`, this command will build the Navigation Simulator. If it compiles
+successful, no errors will be returned. Then, run `./jarvis exec simulators/nav` to run the Simulator.
 
+If desired, you can run a fourth terminal for debugging purposes via LCM messages. To do so, make sure you have another
+terminal, and starting in the `mrover-workspace` directory, run `vagrant ssh`. Once we are ssh'ed into the virtual
```
Comment (cgiger00) [inline jetson/nav/README.md:188]: i would just link the esw guide to this https://github.com/umrover/embedded-testbench/wiki/LCM-Usage

File: jetson/nav/.clang-format [hunk 11]
```diff
@@ -0,0 +1,66 @@
+# Generated from CLion C/C++ Code Style settings
```
Comment (cgiger00) [inline jetson/nav/.clang-format:1]: interesting, is this only possible w CMake not Meson?

File: jetson/nav/README.md [hunk 12]
```diff
@@ -1,156 +1,203 @@
...
\ No newline at end of file
+
+To run the rover at testing, all that needs to be executing is `./jarvis build jetson/nav`
+and `./jarvis exec jetson/nav` to run our code. This should be done AFTER the Perception and GPS executables and
+necessary programs are running. To start, use the Base Station GUI to create waypoints via dropping. (dropping will
+place a waypoint at the current GPS location of the rover) Additionally, the GUI has an option to create a waypoint by
+directly inputting the GPS information Then, drag the newly created waypoint to the course. Repeat this process until
+all desired waypoints are added. Finally, to start the Navigation code, hit the "Autonomy" button on the GUI.
```
Comment (cgiger00) [inline jetson/nav/README.md:203]: can you add a TODO section at the end of this readme? like implement obstacle, etc?

File: jetson/nav/stateMachine.cpp [hunk 13]
```diff
@@ -21,21 +17,24 @@ StateMachine::StateMachine(
...
+ * Executes the correct function or sub state machine based on the current state machine to calculate the next state.
+ * In this way the state machine can be seen as a tree, with search and gate search being subtrees.
+ */
 void StateMachine::run() {
     mPrevTimePoint = mTimePoint;
     auto now = std::chrono::high_resolution_clock::now();
     mTimePoint = now;
 
     // Diagnostic info: take average loop time
+    // TODO move static to instance variable. Fine at the moment since we don't have multiple instances of the state machine
```
Comment (cgiger00) [inline jetson/nav/stateMachine.cpp:37]: can add to readme

**Review by cgiger00 — CHANGES_REQUESTED** [review]
added some comments

File: jetson/nav/README.md (hunk 3 above)
Comment (qhdwight) [reply jetson/nav/README.md:110]: @zwgold made the diagram so I would follow up with him about that

File: jetson/nav/.clang-format (hunk 11 above)
Comment (qhdwight) [reply jetson/nav/.clang-format:1]: No, this is for IDEs. Both VSCode, CLion, VS, etc. support it. It is just a code style guide. Ideally we could lint with it in the future.

File: jetson/nav/README.md (hunk 12 above)
Comment (qhdwight) [reply jetson/nav/README.md:203]: yeah

File: jetson/nav/README.md (hunk 10 above)
Comment (qhdwight) [reply jetson/nav/README.md:188]: kk good to consolidate resources

File: jetson/nav/README.md (hunk 9 above)
Comment (qhdwight) [reply jetson/nav/README.md:160]: I'm not so sure about publishers/subscribers in general tbh

File: jetson/nav/README.md (hunk 3 above)
Comment (zwgold) [reply jetson/nav/README.md:113]: @qhdwight Diagram should be updated now with the changes above.

File: jetson/nav/.clang-format (hunk 11 above)
Comment (eli-schwartz) [reply jetson/nav/.clang-format:1]: As far as cmake / meson goes, Meson at least will notice you have this file and create a target for you to use it: https://mesonbuild.com/Code-formatting.html#clangformat

Comment (qhdwight) [reply jetson/nav/.clang-format:1]: Interesting.

Also... out of curiosity... how did you even find this to comment on. Lol

Comment (niwhsa9) [reply jetson/nav/.clang-format:1]: @eli-schwartz have a read! https://www.rojtberg.net/1481/do-not-use-meson/

File: jetson/nav/README.md (hunk 9 above)
Comment (cgiger00) [reply jetson/nav/README.md:163]: i see

File: jetson/nav/README.md (hunk 12 above)
Comment (cgiger00) [reply jetson/nav/README.md:204]: like with the [x] format

File: jetson/nav/rover.hpp [hunk 14]
```diff
@@ -100,30 +97,28 @@ class Rover {
...
 
-    // A reference to the configuration file.
     const rapidjson::Document& mConfig;
 
-    // A reference to the lcm object that will be used for
-    // communicating with the actual rover and the base station.
+    // Interface for communicating messages to and from the base station
     lcm::LCM& mLcmObject;
 
+<<<<<<< HEAD
```
Comment (qhdwight) [inline jetson/nav/rover.hpp:105]: bruh

**Review by niwhsa9 — APPROVED** [review]
There's no peeposhy on GitHub so you'll have to settle for a LGTM

**Comment by qhdwight** [comment]
@cgiger00 This is ready to go, I think the only tiny thing that is wrong are publishers/subscribers for certain things

**Comment by cgiger00** [comment]
kk sg

File: jetson/nav/stateMachine.cpp [hunk 15]
```diff
@@ -175,62 +169,54 @@ NavState StateMachine::executeDrive() {
...
-}
+}// setSearcher()
 
 void StateMachine::setGateSearcher() {
     mGateStateMachine = GateFactory(weak_from_this(), mConfig);
-}
-
-// Allows outside objects to set the original obstacle angle
-// This will allow the variable to be set before the rover turns
-void StateMachine::updateObstacleDistance(double distance) {
```
Comment (cgiger00) [inline jetson/nav/stateMachine.cpp:214]: was this supposed to be deleted?

**Review by cgiger00 — APPROVED** [review]
pending above thing is fine

File: jetson/nav/stateMachine.cpp (hunk 15 above)
Comment (qhdwight) [reply jetson/nav/stateMachine.cpp:214]: yeah I suppose it could. it's unused

**Pull request merged**